packages/amplify-codegen-model-plugin/src/visitors/appsync-visitor.ts

```diff
@@ -0,0 +1,379 @@
+      return selectedModel ? { [this._parsedConfig.selectedType]: selectedModel } : {};
+    }
+    return this.enumMap;
+  }
+  protected selectedTypeIsEnum() {
+    if (this._parsedConfig && this._parsedConfig.selectedType) {
+      if (this._parsedConfig.selectedType in this.enumMap) {
+        return true;
+      }
+    }
+    return false;
+  }
+
+  /**
+   * returns the Java type or class name
+   * @param field
+   */
+  protected getNativeType(field: CodeGenField): string {
+    const typeName = field.type;
+    let typeNameStr: string = '';
+    if (typeName in this.scalars) {
+      typeNameStr = this.scalars[typeName];
+    } else if (this.isModelType(field)) {
+      typeNameStr = this.getModelName(this.typeMap[typeName]);
+    } else if (this.isEnumType(field)) {
+      typeNameStr = this.getEnumName(this.enumMap[typeName]);
+    } else {
+      throw new Error(`Unknown type ${typeName} for field ${field.name}`);
+    }
+
+    return field.isList ? this.getListType(typeNameStr) : typeNameStr;
+  }
+
+  protected getListType(typeStr: string): string {
+    return `List<${typeStr}>`;
+  }
+
+  protected getFieldName(field: CodeGenField): string {
+    return camelCase(field.name);
+  }
+
+  protected getEnumName(enumField: CodeGenEnum | string): string {
+    if (typeof enumField === 'string') {
+      return pascalCase(enumField);
+    }
+    return pascalCase(enumField.name);
+  }
+
+  protected getModelName(model: CodeGenModel) {
+    return pascalCase(model.name);
+  }
+
+  protected getEnumValue(value: string): string {
+    return upperCase(value);
+  }
+
+  protected isEnumType(field: CodeGenField): boolean {
+    const typeName = field.type;
+    return typeName in this.enumMap;
+  }
+
+  protected isModelType(field: CodeGenField): boolean {
+    const typeName = field.type;
+    return typeName in this.typeMap;
+  }
+
+  protected computeVersion(): string {
+    // Sort types
+    const typeArr: any[] = [];
+    Object.values(this.typeMap).forEach((obj: CodeGenModel) => {
+      // include only key directive as we don't care about others for versioning
+      const directives = obj.directives.filter(dir => dir.name === 'key');
+      const fields = obj.fields
+        .map((field: CodeGenField) => {
+          // include only connection field and type
+          const fieldDirectives = field.directives.filter(field => field.name === 'connection');
+          return {
+            name: field.name,
+            directives: fieldDirectives,
+            type: field.type,
+          };
+        })
+        .sort((a, b) => sortFields(a, b));
+      typeArr.push({
+        name: obj.name,
+        directives,
+        fields,
+      });
+    });
+    typeArr.sort(sortFields);
+    return crypto
+      .createHash('MD5')
+      .update(JSON.stringify(typeArr))
+      .digest()
+      .toString('hex');
+  }
+
+  protected ensureIdField(model: CodeGenModel) {
+    const idField = model.fields.find(field => field.name === 'id');
```

Comment:
This will break when ```@key``` fixes will be implemented as ```@key``` can make any field as ```id``` field.

Reply:
This was an explicit decision. Data Store wants to maintain the hash key and won't let customers change the type or name or value of it. This is needed for conflict resolution